README.md
# GLOW
The GLobal airglOW Model

This directory contains:
   Fortran-90 source code files,
   Makefiles,
   Example input and output files,
   Example job script,
   Subdirectory data/ contains input data files,
   Subdirectory data/iri90 contains IRI input data files


Releasenotes.txt
GLOW version 0.981 release notes, Stan Solomon, 6/2017

Very minor update to v 0.98, implemented to facilitate using GNU gfortran:
   Changed all open statements to use action='read' instead of 'readonly'
   Commented out two lines from nrlmsise00.f; changed 0 to 0.
   Replaced sind and cosd functions in snoem.f90, and general cleanup.
   Corrected rcolum.f90 to call chap with pi-chi when chi>pi/2.
   Added a trap to etrans.f90 to assure no calculation when j=1.
   Zeroed out all arrays in cglow.f90.
   None of these changes should have any effect on results.

-----------------------------------------------------

GLOW version 0.98 release notes, Stan Solomon, 3/2017

See Quickstart.txt for very basic instructions
See Glow.txt for more details
See Glowlicense.txt for the Open Source Academic Research License Agreement

This is now a Fortran-90 code, with MPI and netCDF options
  No more common blocks.
  No more header file.
  Use-associated variables are defined in module cglow.f90
  Most code converted to lower-case free-form style
Example driver program glowbasic provides functionality in previous versions
  Can run one profile at a time or loop
  Single-processor, no MPI required
  No netCDF required either
  You can easily modify this program to suit particular applications
New driver program glowdriver uses TIE-GCM, TIME-GCM or empirical model inputs
  Ultimately this will include WACCM-X as well
  Fully MPI code with user-specified number of processors
  Namelist input from file in.namelist
  netCDF output to a file specified in the namelist input
Can still run with empirical models:
  Default empirical neutral model updated to MSIS-2K
  NOEM empirical nitric oxide model, based on SNOE data, used for empirical runs
  IRI-90 still used for empirical ionosphere

New features in v. 0.98:
  Most of the new things in v. 0.98 are computational, not scientific.
  Move to MPI parallel processing forced the elimination of common blocks.
  Reads netCDF input files from GCMs, and writes global netCDF output files.
  All of this is optional - you can still run the model the "old" way.
    (But old driver programs will have to be updated to use shared variables.)
  Empirical model calls are consolidated in subroutine MZGRID
  Calculates conductivities using Ryan McGranaghan's CONDUCT routine (adapted).
  N2 LBH, O 1356, N 1493, O 1304 added to airglow output (ZETA array).
    (Note 1493 is very approximate since branching rations are poorly known.)
  Radiative recombination and mutual neutralization included in 1356.
  Subroutine BANDS added to allocate LBH among v' levels.
    (Really just a stub now, will ultimately handle vibrational distributions.)

Note that several bugfixes and parameter adjustments were made following the
  "beta" release of v. 0.98 in early January 2017, so do not use that version.

Other significant changes:

   The quartic equation solver formerly used by GCHEM to calculate electron
density has been abandoned because it produced minor errors and could not
be verified.  Instead, a simple iterative method is used, which turns out
to be more accurate and equally fast.  Thanks to Michael Hirsch for identifying
this problem.

   The high-energy electron precipitation example (hexexample) is no longer
provided.  This functionality is built-in to the codes, but if you want to
do a high-energy electron run you will have to increase NBINS in cglow.f90
(e.g., to 343 for energies up to 100 MeV) and extend the altitude grid to
lower altitude by increasing JMAX in the driver program, and adding levels
in mxgrid.f90.

   Conjugate hemisphere photoelectrons are no longer provided in the example
drivers, since they are very minor, triple the run time, and weren't widely
used.  They can easily be re-installed in a driver program, but note that
the magnetic field needs to be updated.  The calculation is fairly approximate
anyway because it doesn't consider exosphere/plasmasphere attenuation of
the electron flux.

   The background ("night time") ionization rate calculated in QBACK was updated
to conform to the forumlation now used in the TIE-GCM, TIME-GCM, WACCM, and
WACCM-X.  It is still pretty approximate, but produces an night ionosphere in
the E and F1 regions that is in reasonable agreement with observations.

   Several unused artifacts were eliminated, but arrays PIA and SESPEC were
retained in anticipation of including proton aurora ionization in the future.

Known issues (most of these carried over from v. 0.97):
  X-rays shortward of 18 A need to be re-examined and updated.
  Magnetic field (GEOMAG, FIELDM) is out of date.  GEOMAG is only used by
    the NOEM model, and isn't too different from the coordinate transform
    used to generate SNOE data on a magnetic grid.  FIELDM only calculates
    the dip angle, which has a minimual effect on results.  But these should
    be updated to time-varying IGRF/Apex coordinates.
  Temporary Y2K fix to SUNCOR.  Also should update to J2000 epoch.
    Should be fine for 1950-2050, but really need to make if valid for either
    yyddd to yyyyddd date format, which will extend the range of validity to
    1900-2100.  (This only affects SZA.)
  O(1S) needs to be re-evaluated (still).  Nightglow recombination source
    still not included.  Also need to include O2 A-bands.
  O(1S) from O2 dissociation (BSO2) is currently hardwired - only works with
    LMAX=123 (although changes shortward of 800 A are OK).
  IRI should be updated.  Usually, IRI is Only used for electron density above
    200 km, and IRI hasn't changed much in the F-region, so it's OK for now.
  Cascade contributions to 7774 and subsequent cascade to 1356 is questionable.
    I reduced the effective 7774 cross section, so 7774 and 1356 are now
    self-consistent, and in reasonable agreement with GUVI data.  
  1493 sources and branching ratios are speculative.  For now, I am presuming
    that it is produced during photodissociative ionization and electron

glow.f90
! Subroutine GLOW

! This software is part of the GLOW model.  Use is governed by the Open Source
! Academic Research License Agreement contained in the file glowlicense.txt.
! For more information see the file glow.txt.

! Version 0.98, 1/2017
! Version 0.981, 6/2017

! Stan Solomon, 1988, 1989, 1990, 1991, 1992, 1994, 2000, 2002, 2005, 2015, 2016
!
! Subroutine GLOW is the master routine of the /glow package.
! It receives input parameters from the calling program using use-associated variables
! defined in module CGLOW, calls the other subroutines, and returns results.
! CGLOW also defines the array dimensions for the altitude grid, the electron
! energy grid, solar spectrum wavelengh bins, etc.

! Call subroutine CGLOW_INIT first to set array dimensions and allocate variables
! Call subroutine EGRID before call to GLOW to set up electron energy grid
! Call subroutine MAXT before call to GLOW to specify auroral electron flux (if any)

! Subroutines called by GLOW are:
!   FIELDM  calculates magnetic dip angle
!   SOLZEN  calculates solar zenith angle
!   SSFLUX  scales solar flux for activity level
!   RCOLUM  calculates slant column density of major species
!   EPHOTO  calculates photoionization and photoelectron production
!   QBACK   estimates background ('night time') ionization
!   ETRANS  computes electron transport, ionization, excitation
!           calls EXSECT for cross-sections, first call only
!   GCHEM   finds electron/ion/metastable densities, airglow emissions
!   BANDS   calculates vibrational distributions for selected band systems (currently only LBH)

! Supplied to subroutine using use-associated data defined in module CGLOW:
! IDATE   Date, in form yyddd
! UT      Universal Time; seconds
! GLAT    Geographic latitude; degrees
! GLONG   Geographic longitude; degrees
! ISCALE  Solar flux scaling switch, see subroutine SSFLUX
! JLOCAL  =0 for electron transport calculation, =1 for local calc only
! KCHEM   Ion/electron chemistry switch, see subroutine GCHEM
! F107    Solar 10.7 cm flux for day being modeled, 1.E-22 W m-2 Hz-1
! F107A   Solar 10.7 cm flux 81-day centered average
! XUVFAC  Factor by which to multiply to solar flux 16-250 A or 16-50 A.
! ZZ      altitude array; cm
! ZO      O number density at each altitude; cm-3
! ZN2     N2  "      "      "   "     "       "
! ZO2     O2         "
! ZNO     NO         "
! ZNS     N(4S)      "
! ZND     N(2D)      "
! ZRHO    mass density at each altitude; gm cm-3 (not currently in use)
! ZE      electron density at each alt; cm-3
! ZTN     neutral temperature at each alt; K
! ZTI     ion temperature at each alt; K
! ZTE     electron temp at each alt; K
! PHITOP  energetic electron flux into top of atmosphere; cm-2 s-1 eV-1

! Calculated by subroutine and returned using use-associated data defined in module CGLOW:
! SZA     solar zenith angle; radians
! DIP     magnetic field dip angle; radians
! EFRAC   energy conservation check from ETRANS, (out-in)/in
! IERR    error code returned from ETRANS:
!           0=normal, 1=local problem, 2=transport problem
! ZMAJ    major species density array, O, O2, N2; cm-3
! ZCOL    major species slant column density array, O, O2, N2; cm-2
! WAVE1   longwave edge of solar flux wavelength range; A
! WAVE2   shortwave edge of solar flux wavelength range; A
! SFLUX   scaled solar flux in each wavelength range; photons cm-2 s-1
! ENER    electron energy grid; eV
! DEL     width of each bin in electron energy grid; eV
! PESPEC  photoelectron production rate at energy, altitude; cm-3 s-1
! PIA     proton aurora ionization rate (not currently in use); cm-3 s-1.
! SESPEC  proton aurora secondary electron production rate (not currently in use); cm-3 s-1
! PHOTOI  photoionization rates for state, species, altitude; cm-3 s-1
!           O+ states: 4S, 2Do, 2Po, 4Pe, 2Pe
!           O2+ states: X, a+A, b, dissoc.
!           N2+ states: X, A, B, C, F, dissoc.
! PHOTOD  photodissoc. & exc. rates for state, species, alt.; cm-3 s-1
!           (1,2,J) = O2 -> O(3P) + O(1D))
!           (2,2,J) = O2 -> O(3P) + O(1S)
!           (1,3,J) = N2 -> N + N
! PHONO   photoionization/dissociation/excitation rates for NO, cm-3 s-1
!         (1,J) = NO+ from H Ly-a ionization
! SION    electron impact ionization rates calculated by ETRANS; cm-3 s-1
! UFLX    upward hemispherical electron flux; cm-2 s-1 eV-1
! DFLX    downward hemispherical electron flux; cm-2 s-1 eV-1
! AGLW    Electron impact exc. rates; state, species, alt.; cm-3 s-1
!           O states: 1D, 1S, 5S, 3S, 3p5P, 3p3P, 3d3D, 3s'3D
!           O2 states: a, b, (A+A'+c), B(SRC), 9.9eV, Ryds., vib.
!           N2 states: (A+B+W), B', C, (a+a'+w), 1Pu, b', Ryds., vib.
! EHEAT   ambient electron heating rate, eV cm-3 s-1
! TEZ     total energetic electron energy deposition, eV cm-3 s-1
! TPI     total photoionization rate at each altitude, cm-3 s-1
! TEI     total electron impact ionization rate at each altitude, cm-3 s-1
! TIR     total ionization rate at each altitude (TPI+TEI), cm-3 s-1
! ECALC   electron density, calculated below 200 km, cm-3
! ZXDEN   array of excited and and/or ionized state densities at each altitude:
!           O+(2P), O+(2D), O+(4S), N+, N2+, O2+, NO+, N2(A), N(2P),
!           N(2D), O(1S), O(1D); cm-3
! ZETA    array of volume emission rates at each altitude:
!           3371A, 4278A, 5200A, 5577A, 6300A, 7320A, 10400A, 3466A,
!           7774A, 8446A, 3726A, LBH, 1356, 1493, 1304; cm-3 s-1
! ZCETA   array of contributions to each v.e.r at each alt; cm-3 s-1
! VCB     array of vertical column brightnesses (as above); Rayleighs

! Array dimensions:
! JMAX    number of altitude levels
! NBINS   number of energetic electron energy bins
! LMAX    number of wavelength intervals for solar flux
! NMAJ    number of major species
! NST     number of states produced by photoionization/dissociation
! NEI     number of states produced by electron impact
! NEX     number of ionized/excited species
! NW      number of airglow emission wavelengths
! NC      number of component production terms for each emission


    subroutine glow

      use cglow,only: jmax,lmax,nw,nst,nmaj,nbins,iscale,nei,ierr, &
                      glat,glong,idate,ut,f107,f107a, &
                      ener,del,dip,sza,xuvfac, &
                      wave1,wave2,sflux,zmaj,zo,zo2,zn2,zz,ztn,zcol, &
                      photoi,photod, phono,pespec,pia,sespec,phitop, &
                      uflx,dflx,sion,aglw,eheat,tez, efrac,zno

      implicit none

      real :: zvcd(nmaj,jmax),xf,yf,zf,ff,dec,sdip,teflux

      real,parameter :: pi=3.1415926536
      integer,save :: ifirst=1
      integer :: j,i,ist,n,iei

! First call only: set up energy grid:

      if (ifirst == 1) then
        ifirst = 0
        call egrid (ener, del, nbins)
      endif

! Find magnetic dip angle and solar zenith angle (radians):

      call fieldm (glat, glong, 300., xf, yf, zf, ff, dip, dec, sdip)
      dip = abs(dip) * pi/180.
      if (dip < 0.01) dip=0.01

      call solzen (idate, ut, glat, glong, sza)
      sza = sza * pi/180.

! Scale solar flux:

      call ssflux (iscale, f107, f107a, xuvfac, wave1, wave2, sflux)

! Pack major species density array:

      do j=1,jmax
        zmaj(1,j) = zo(j)
        zmaj(2,j) = zo2(j)
        zmaj(3,j) = zn2(j)
      enddo

! Calculate slant path column densities of major species in the
! direction of the sun:

      call rcolum (sza, zz, zmaj, ztn, zcol, zvcd, jmax, nmaj)

! Call subroutine EPHOTO to calculate the photoelectron production
! spectrum and photoionization rates as a function of altitude,
! unless all altitudes are dark, in which case zero arrays:

      if (sza < 1.85) then
        call ephoto
      else
        photoi(:,:,:) = 0.0
        photod(:,:,:) = 0.0
        phono(:,:) = 0.0
        pespec(:,:) = 0.0
      endif

! Zero proton aurora ionization rates and secondary electron production (not currently in use):

      pia(:,:) = 0.0
      sespec(:,:) = 0.0

! Add background ionization to photoionization:

      call qback (zmaj, zno, zvcd, photoi, phono, f107, jmax, nmaj, nst)

! Call subroutine ETRANS to calculate photoelectron and auroral
! electron transport and electron impact excitation rates, unless
! there are no energetic electrons, in which case zero arrays:

      teflux = 0.
      do n=1,nbins
        teflux = teflux + phitop(n)
      enddo

      if (teflux > 0.001 .or. sza < 1.85) then
        call etrans
      else
        uflx(:,:) = 0.0
        dflx(:,:) = 0.0
        sion(:,:) = 0.0
        aglw(:,:,:) = 0.0
        eheat(:) = 0.0
        tez(:) = 0.0
        efrac = 0.0
        ierr = 0
      endif

! Call subroutine GCHEM to calculate the densities of excited and
! ionized consituents, airglow emission rates, and vertical column
! brightnesses:

      call gchem

! Call subroutine BANDS to calculate band-specific airglow emission
! rates (currently only LBH upper state distribution):

      call bands

      return

    end subroutine glow
